Transition from guest-get-fsinfo qemu guest agent command to libvirt api

Ruclo opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Removes the qemu guest agent guest-get-fsinfo command execution and uses libvirt api instead to retrieve information about filesystems with the goal of unifying the API.

### References
<!-- optional,
  Use `Fixes #<issue number>(, Fixes #<issue_number>, ...)` format, to close the issue(s) when PR gets merged.
  Use `Partially addresses #<issue number>` to link an issue without closing it when the PR merges.
-->

- Fixes [CNV-54838](https://issues.redhat.com/browse/CNV-54838)
- Partially addresses #13651 [CNV-54839](https://issues.redhat.com/browse/CNV-54839)

<!-- optional,
  VEP tracking issue if this PR is implementing one.
  For additional info about VEP tracking issue, see https://github.com/kubevirt/enhancements#process
  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issues/<vep_tracking_issue_number>
-->


### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

Ruclo commented:
/cc @akrejcir

akrejcir commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -327,6 +316,10 @@ func CreatePoller(
 			},
+			{
+				CallTick:  qemuAgentFileInterval,
+				InfoTypes: libvirt.DOMAIN_GUEST_INFO_FILESYSTEM | libvirt.DOMAIN_GUEST_INFO_DISKS,
```
Please put the second constant on a new line, so it looks similar to the above definition.

akrejcir commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -588,6 +574,47 @@ func convertToUsers(guestInfo *libvirt.DomainGuestInfo) []api.User {
 }
 
+func convertToDisks(guestInfo *libvirt.DomainGuestInfo, fileSystem libvirt.DomainGuestInfoFileSystem) []api.FSDisk {
+	var disks []api.FSDisk
```
Please define variable close to where it is used. The code is then easier to read.

akrejcir commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -588,6 +574,47 @@ func convertToUsers(guestInfo *libvirt.DomainGuestInfo) []api.User {
+	for _, disk := range fileSystem.Disks {
+		diskInfo := findDisk(disk.Serial)
+		disks = append(disks, api.FSDisk{
+			Serial:  diskInfo.Serial,
```
I think using `disk.Serial` here makes the intention clearer, even if the 2 serials are the same.

Ruclo replied:
I like that both come from the same struct. It's more readable imo

akrejcir replied:
Ok

akrejcir commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -588,6 +574,47 @@ func convertToUsers(guestInfo *libvirt.DomainGuestInfo) []api.User {
+			}
+		}
+		// Unreachable
+		return libvirt.DomainGuestInfoDisk{}
```
This code could be reachable, because it depends on what values libvirt sends. 
In case the serial does not exist, the bus type can be `""`.

Ruclo replied:
Returning the zero value causes the bus type to be empty string

akrejcir commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -2832,8 +2797,8 @@ type VirtualMachineInstanceFileSystem struct {
-	UsedBytes      int                                    `json:"usedBytes"`
-	TotalBytes     int                                    `json:"totalBytes"`
+	UsedBytes      uint64                                 `json:"usedBytes"`
+	TotalBytes     uint64                                 `json:"totalBytes"`
```
Same here.

akrejcir reviewed (commented):
Functionality looks good. I mostly have code style comments.

Ruclo commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -588,6 +574,47 @@ func convertToUsers(guestInfo *libvirt.DomainGuestInfo) []api.User {
+	return disks
+}
+
+func convertToFilesystems(guestInfo *libvirt.DomainGuestInfo) []api.Filesystem {
```
I kept it aligned with the other functions. I think the semantics mean converting DomainGuestInfo struct to api structs. Perhaps extract would be more suitable verb instead of convert?

akrejcir reviewed (commented):
I think you've committed the second merge commit by mistake. Can you fix it?

Ruclo commented:
I decided to use the libvirt constants, but kept the key for filesystem and disk info merged as the AgentStore stores api objects and I could either store separate incomplete api.Filesystem and api.FSDisk objects and merge them on GetFS or merge them once when retrieving the domain guest info. Perhaps using some aliases for the constants would not be a bad decision in the future?

Ruclo committed:
refactor(api): Move from guest-get-fsinfo ga command to libvirt api

In this commit, the execution of guest-get-fsinfo
guest agent command in agent poller got removed and the information
about filesystems is provided via libvirt api instead,
which is tested better.

Signed-off-by: Michal Vavrinec <mvavrine@redhat.com>

codingben reviewed (changes requested):
Looks okay, need to review it again before I'll add lgtm. Can you please address and or resolve @akrejcir's comments?

akrejcir commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -486,6 +476,10 @@ func fetchAndStoreGuestInfo(infoTypes libvirt.DomainGuestInfoTypes, agentPoller
 		agentPoller.agentStore.Store(libvirt.DOMAIN_GUEST_INFO_USERS, convertToUsers(guestInfo))
 	}
+
+	if infoTypes&FSInfoType != 0 {
```
Using the constant here hides that this condition is different than those above. When the constant is expanded, the condition becomes:
```go
if infoTypes&libvirt.DOMAIN_GUEST_INFO_FILESYSTEM != 0 || infoTypes&libvirt.DOMAIN_GUEST_INFO_DISKS != 0 {}
```
Which is probably not what we want. I think we want both info types to be present, not just one of them.

Ruclo commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -588,6 +574,47 @@ func convertToUsers(guestInfo *libvirt.DomainGuestInfo) []api.User {
 	return users
 }
 
+func convertToDisks(guestInfo *libvirt.DomainGuestInfo, fileSystem libvirt.DomainGuestInfoFileSystem) []api.FSDisk {
```
the other convert functions also take DomainGuestInfo, so I tried to keep it consistent

codingben commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -588,6 +574,47 @@ func convertToUsers(guestInfo *libvirt.DomainGuestInfo) []api.User {
+func convertToDisks(guestInfo *libvirt.DomainGuestInfo, fileSystem libvirt.DomainGuestInfoFileSystem) []api.FSDisk {
+	var disks []api.FSDisk
+
+	findDisk := func(serial string) libvirt.DomainGuestInfoDisk {
```
@Ruclo Can you please address it? Other than that looks good.

codingben replied:
@Ruclo I'm not fan of lambda functions and in the past year we're working to eliminate them from the codebase. Can you please refactor this?

codingben commented:
@Ruclo We're trying to get rid of lambda functions. Can you please address it [here](https://github.com/kubevirt/kubevirt/pull/15984/files#r2554161833)?

/lgtm
/hold

codingben commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -41,10 +41,10 @@ type AgentCommand string
 	GetAgent          AgentCommand = "guest-info"
 	GetFSFreezeStatus AgentCommand = "guest-fsfreeze-status"
 
+	FSInfoKey           = "FSInfo"
```
I'm not sure about this constant, can you just use `libvirt.DOMAIN_GUEST_INFO_FILESYSTEM` instead? I think it would be better to keep it consistent with others keys from libvirt package.

@akrejcir WDYT?

codingben reviewed (changes requested):
Please check the unit tests and refactor the function to be without lambda function inside.

akrejcir commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -588,10 +582,58 @@ func convertToUsers(guestInfo *libvirt.DomainGuestInfo) []api.User {
+	}
+	// Unreachable
+	return libvirt.DomainGuestInfoDisk{}
+}
```
I think a better design here would be to return a pointer to the found disk, and `nil` if it is not found.
That eliminates copying, so has a little better performance, and handles the case when somehow the disk is not found.

akrejcir commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller_test.go:
```diff
@@ -102,6 +104,9 @@ var _ = Describe("Qemu agent poller", func() {
 			Expect(users[0].Name).To(Equal("admin"))
+
+			filesystems := agentStore.GetFS(1)
+			Expect(filesystems[0].Name).To(Equal("tmpfs"))
```
To avoid the potential panic, you can also add:
```golang
Expect(filesystems).To(HaveLen(1))
```

codingben commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -588,10 +582,58 @@ func convertToUsers(guestInfo *libvirt.DomainGuestInfo) []api.User {
+			return disk
+		}
+	}
+	// Unreachable
```
Does "unreachable" means it's a dead-code here? If so, in that case, just have one single return in the end that will return an existing or empty disk.

codingben reviewed (changes requested):
/lgtm

I think from my side it looks good, just one last comment about the unreachable code in your new function.

0xFelix commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -327,6 +323,10 @@ func CreatePoller(
 				InfoTypes: libvirt.DOMAIN_GUEST_INFO_LOAD,
 			},
+			{
+				CallTick:  qemuAgentFileInterval,
```
Could it be added to an existing poller?

0xFelix commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -588,10 +582,58 @@ func convertToUsers(guestInfo *libvirt.DomainGuestInfo) []api.User {
+		}
+	}
+	// Unreachable
+	return libvirt.DomainGuestInfoDisk{}
```
Return ptr from func and nil if disk was not found

0xFelix reviewed (commented):
What's the state of this?

vladikr commented:
/remove-lifecycle rotten

vladikr commented:
any progress here?
